src/main/java/com/flightstats/hub/webhook/LocalWebhookRunner.java

```diff
@@ -87,31 +88,58 @@ private boolean ensureRunningWithLock(String name) {
 
     private boolean start(Webhook daoWebhook) {
         WebhookLeader webhookLeader = webhookLeaderProvider.get();
-        boolean hasLeadership = webhookLeader.tryLeadership(daoWebhook);
-        if (hasLeadership) {
-            localLeaders.put(daoWebhook.getName(), webhookLeader);
-        }
-        return hasLeadership;
+        return webhookLeader.tryLeadership(daoWebhook, localLeaders);
     }
 
     void stopAll() {
-        runAndWait("LocalWebhookRunner.stopAll", localLeaders.keySet(), this::stop);
+        runAndWait("LocalWebhookRunner.stopAll", localLeaders.getLocalLeaders().keySet(), this::stop);
     }
 
     void stop(String name) {
+        lockManager.executeLocked(name, () -> stopWithLock(name));
+    }
+
+    private void stopWithLock(String name) {
         log.info("stopping {} if running local", name);
-        if (localLeaders.containsKey(name)) {
+        WebhookLeader webhookLeader = localLeaders.getLocalLeaders().get(name);
```

Comment:
stylistically, I'd probably wrap the `.getLocalLeaders().get(name)` up in a helper method on LocalLeaders that returns an `Optional<WebhookLeader>`. Then this'd become
```
localLeaders.getLeader(name)
        .peek(leader -> log.info("stopping local {}", name))
        .ifPresent(WebhookLeader::exit);
```

That doesn't help the other case above (`ensureRunningWithLock`) very much because it's kind of a mess of a method, though, so it's not much of a win other than abstracting away the data structure that we're storing the leader data in.